internal/restore/pvc_action.go

```diff
@@ -172,7 +172,48 @@ func (p *PVCRestoreItemAction) Execute(input *velero.RestoreItemActionExecuteInp
 		setPVCStorageResourceRequest(&pvc, restoreSize, p.Log)
 	}
 
-	resetPVCSpec(&pvc, volumeSnapshotName)
+	// Get the Storage Class
+	var csiDriverName string
+	storageClassName := pvc.Spec.StorageClassName
+	if storageClassName != nil {
+		storageClass, err := clientset.StorageV1().StorageClasses().Get(context.TODO(), *storageClassName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, fmt.Sprintf("Failed to get StorageClass %s to check PVC %s/%s provisioner", *storageClassName, pvc.Namespace, pvc.Name))
+		}
+		csiDriverName = storageClass.Provisioner
+	} else {
+		p.Log.Infof("StorageClass is not set for PVC %s/%s", pvc.Namespace, pvc.Name)
+	}
+
+	if csiDriverName != "file.csi.azure.com" {
+		resetPVCSpec(&pvc, volumeSnapshotName)
+	} else {
+		// Set annotation that the PVC users Azure File CSI driver
+		annotations := map[string]string{
+			util.CSIDriverNameAnnotation: "file.csi.azure.com",
+		}
+		util.AddAnnotations(&pvc.ObjectMeta, annotations)
+		p.Log.Infof("Found Azure Files CSI driver. PVC data source will not be changed.")
+	}
+
+	// Add annotation to the PVC with VolumeSnapshotContent name
+	vscList, err := snapClient.SnapshotV1beta1().VolumeSnapshotContents().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, fmt.Sprintf("Failed to list VolumeSnapshotContents to restore PVC %s/%s", pvc.Namespace, pvc.Name))
+	}
+	var volumeSnapshotContentName string
+	for _, vsc := range vscList.Items {
+		if vsc.Spec.VolumeSnapshotRef.Name == vs.Name {
+			volumeSnapshotContentName = vsc.Name
+		}
+	}
+	if volumeSnapshotContentName == "" {
+		return nil, fmt.Errorf("Failed to get VolumeSnapshotContent for VolumeSnapshot %s/%s", vs.Namespace, vs.Name)
+	}
+	vscAnnotations := map[string]string{
+		"volume-snapshot-content-name": volumeSnapshotContentName,
+	}
+	util.AddAnnotations(&pvc.ObjectMeta, vscAnnotations)
```

Comment:
The VolumeSnapshotContent will be removed if the whole namespace is gone, and the Azure Files restore will not work. This needs to be changed.